teps/0040-ignore-step-errors.md

```diff
@@ -0,0 +1,157 @@
+    "reason": "Failed",
+    "status": "False",
+    "type": "Succeeded"
+  }
+]
+```
+
+If such a task with a failing step is part of a pipeline, the `pipelineRun` stops executing and subsequent steps in that
+task (similar to `taskRun`) stop executing any other task in the pipeline which results in a pipeline failure.
+
+```yaml
+$ kubectl get pr pipelinerun-with-failing-step-csmjr -o json | jq .status.conditions
+[
+  {
+    "lastTransitionTime": "2021-02-02T18:51:15Z",
+    "message": "Tasks Completed: 1 (Failed: 1, Cancelled 0), Skipped: 3",
+    "reason": "Failed",
+    "status": "False",
+    "type": "Succeeded"
+  }
+]
+```
+
+Many common tasks have the requirement where a step failure must not stop executing the rest of the steps.
+In order to continue executing subsequent steps, task authors have the flexibility of wrapping an image and
+exiting that step with success. This changes the failing step into a success and does not block further
+execution. But, this is a workaround and only works with images that can be wrapped:
+
+```shell
+    steps:
+    - image: docker.io/library/golang:latest
+      name: ignore-unit-test-failure
+      script: |
+        go test .
+        TEST_EXIT_CODE=$?
+        if [ $TEST_EXIT_CODE != 0 ]; then
+          exit 0
+        fi
+```
+
+This workaround does not apply to off-the-shelf container images.
+
+As a pipeline execution engine, we want to support off-the-shelf container images as a step, and provide
+the option to ignore such step errors. The task author can choose to continue execution, capture the original non-zero
+exit code, and make it available for the rest of the steps in that task.
+
+Issue: [tektoncd/pipeline#2800](https://github.com/tektoncd/pipeline/issues/2800)
+
+
+## Motivation
+
+It should be possible to easily use off-the-shelf (OTS) images as steps in Tekton tasks. A task author has no
+control over the image but may desire to ignore an error and continue executing the rest of the steps.
+
+
+### Goals
+
+Design a step failure strategy so that the task author can control the behaviour of the underlying step and decide
+whether to continue executing the rest of the steps in the event of failure.
+
+Store the step container's termination state and make it accessible to the rest of the steps in a task.
+
+Be applicable to any container image including custom or off-the-shelf images.
+
+### Non-Goals
+
+This proposal is limited to a step within a task and does not address `pipelineTask` level failure case.
+
+## Requirements
+
+* Users should be able to use prebuilt images as-is without having to understand if a shell or similar capability exists
+  in an image and then altering the entrypoint to allow capturing errors.
+
+* It should be possible to know that a step failed and subsequent steps allowed to continue by observing the status of
+  the `TaskRun`.
```

Comment:
I'd also like to include a requirement that other steps within the Task can access this information, I believe @afrittoli disagrees - I'm hoping we can include this as a requirement but also could deliver the feature incrementally such that the initial implementation doesn't need to support this.

I feel strongly that we haven't sufficiently addressed the use cases and problem statement without it

Reply:
I have added that (other steps within the task can access the process exit code) in the requirements. @afrittoli please let me know if you still disagree or have any other ideas.